File: solutions/ruby/resistor-color-duo/1/resistor_color_duo.rb
module ResistorColor

  def self.colors
    %w[black brown red orange yellow green blue violet grey white]
  end

  COLORS = colors

  BAND = {
    black:  0,
    brown:  1,
    red:    2,
    orange: 3,
    yellow: 4,
    green:  5,
    blue:   6,
    violet: 7,
    grey:   8,
    white:  9
  }

  def self.color_code(color)
    BAND[color.to_sym]
  end

end

#require_relative 'resistor_color'

module ResistorColorDuo

  def self.value(colors)
    digits = colors.first(2).map { |color|
      ResistorColor::BAND[color.to_sym]
    }

    digits.first * 10 + digits.last
  end

end

File: solutions/elixir/resistor-color-duo/3/lib/resistor_color_duo.ex
defmodule ResistorColorDuo do

  @colors_with_digits Enum.with_index(
    [:black,
     :brown,
     :red,
     :orange,
     :yellow,
     :green,
     :blue,
     :violet,
     :grey,
     :white
     ]
                                       )
    # Nested comprehension: generates function clauses for ALL 100 combinations
    # (10 colors × 10 colors = 100 clauses)
    for {color1, digit1} <- @colors_with_digits,
        {color2, digit2} <- @colors_with_digits do
    # unquote injects the actual color atom (e.g., :black) into the function head
    # unquote injects the actual digit (e.g., 0) into the function body
    def value([unquote(color1), unquote(color2) | _]) do
      unquote(digit1) * 10 + unquote(digit2)
    end
  end

end
